core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1571,4 +1571,20 @@ public void onAfter() {
         return false;
     }
 
+    /**
+     * Simple struct encapsulating a shard failure
+     * @see IndexShard#addShardFailureCallback(Callback)
+     */
+    public static final class ShardFailure {
+        public final ShardRouting routing;
+        public final String reason;
+        public final Throwable cause;
```

Comment:
We use @nullable everywhere (including in `onFailedEngine`) to indicate that something can be null and people should watch out for it. I personally like it but if we end up not liking it, we should just kill `@nullable` and all it's usages. If we want to remove it, maybe open and issue and see what people think?